internal/components/setup/compose.go

```diff
@@ -17,3 +17,212 @@
 //
 
 package setup
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"regexp"
+	"syscall"
+	"time"
+
+	"github.com/apache/skywalking-infra-e2e/internal/config"
+	"github.com/apache/skywalking-infra-e2e/internal/constant"
+	"github.com/apache/skywalking-infra-e2e/internal/logger"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/client"
+
+	"github.com/testcontainers/testcontainers-go"
+)
+
+// ComposeSetup sets up environment according to e2e.yaml.
+func ComposeSetup(e2eConfig *config.E2EConfig) error {
+	composeConfigPath := e2eConfig.Setup.GetFile()
+	if composeConfigPath == "" {
+		return fmt.Errorf("no compose config file was provided")
+	}
+
+	// build docker client
+	cli, err := client.NewClientWithOpts(client.FromEnv)
+	if err != nil {
+		return err
+	}
+
+	// setup docker compose
+	composeFilePaths := []string{
+		composeConfigPath,
+	}
+	identifier := GetIdentity()
+	compose := testcontainers.NewLocalDockerCompose(composeFilePaths, identifier)
+	execError := compose.WithCommand([]string{"up", "-d"}).Invoke()
+	if execError.Error != nil {
+		return execError.Error
+	}
+
+	// record time now
+	timeNow := time.Now()
+	timeout := e2eConfig.Setup.Timeout
+	var waitTimeout time.Duration
+	if timeout <= 0 {
+		waitTimeout = constant.DefaultWaitTimeout
+	} else {
+		waitTimeout = time.Duration(timeout) * time.Second
+	}
+	logger.Log.Debugf("wait timeout is %d seconds", int(waitTimeout.Seconds()))
+
+	// find exported port and build env
+	for service, content := range compose.Services {
+		serviceConfig := content.(map[interface{}]interface{})
+		ports := serviceConfig["ports"]
+		if ports == nil {
+			continue
+		}
+		portList := ports.([]interface{})
+		container, err := findContainer(cli, fmt.Sprintf("%s_%s", identifier, getInstanceName(service)))
+		if err != nil {
+			return err
+		}
+		containerPorts := container.Ports
+
+		for inx := range portList {
+			for _, containerPort := range containerPorts {
+				if int(containerPort.PrivatePort) != portList[inx].(int) {
+					continue
+				}
+
+				// calculate max wait time
+				waitTimeout = NewTimeout(timeNow, waitTimeout)
+				timeNow = time.Now()
+
+				// wait port and export
+				err := waitTCPPortStarted(context.Background(), cli, container, int(containerPort.PublicPort), int(containerPort.PrivatePort), waitTimeout)
+				if err != nil {
+					return fmt.Errorf("could wait port exported: %s:%d, %v", service, portList[inx], err)
+				}
+
+				// expose env config to env
+				envKey := fmt.Sprintf("e2e_host_and_port_%s_%d", service, containerPort.PrivatePort)
```

Comment:
Let's only expose the port as env var `<SERVICE_NAME>_PORT`, because the host is not necessarily always 127.0.0.1, so that user can use something like `localhost:${WEBAPP_PORT}` as the url of trigger